File: leetcode/392_is_subsequence.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):
    def isSubsequence(self, s, t):
        if not s:
            return True

        i = 0
        for c in t:
            if i < len(s) and c == s[i]:
                i += 1

        return i == len(s)


if __name__ == '__main__':
    solution = Solution()

    assert solution.isSubsequence('abc', 'ahbgdc')
    assert not solution.isSubsequence('axc', 'ahbgdc')
    assert solution.isSubsequence('', 'ahbgdc')
    assert solution.isSubsequence('a', 'ab')

File: leetcode/392_is_subsequence.cpp
#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
    bool isSubsequence(string s, string t) {
        if (s.empty()) {
            return true;
        }

        int i = 0, l = s.length();
        for (auto c : t) {
            if (c == s[i]) {
                i++;
                if (i == l) {
                    return true;
                }
            }
        }

        return false;
    }
};

int main() {
    auto solution = Solution();

    assert(solution.isSubsequence("abc", "ahbgdc"));
    assert(!solution.isSubsequence("axc", "ahbgdc"));
    assert(solution.isSubsequence("", "ahbgdc"));
    assert(solution.isSubsequence("a", "ab"));
}
